### PR TITLE
stricter type checking for task terminal reconnection

### LINKED ISSUE #181102: ERR Cannot read properties of undefined (reading 'lastTask')
Saw this in insiders after a window reload:

```
ERR Cannot read properties of undefined (reading 'lastTask'): TypeError: Cannot read properties of undefined (reading 'lastTask')
    at P.Bb (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1516:17006)
    at P.Cb (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1516:17101)
    at vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1516:20463
    at async P.Fb (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1516:20483)
    at async P.xb (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/workbench/workbench.desktop.main.js:1516:8701)
```

Seems to be from:

image

### PR DESCRIPTION
**Pull request by meganrogge**
fixes https://github.com/microsoft/vscode/issues/181102

### CODE REVIEW

**Review by connor4312 — DISMISSED** [review]

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 1]
```diff
@@ -1323,18 +1326,21 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
...
 			return;
 		}
-		this._reconnectedTerminals = this._terminalService.getReconnectedTerminals(ReconnectionType)?.filter(t => !t.isDisposed);
+		this._reconnectedTerminals = [];
+		for (const terminal of this._terminalService.getReconnectedTerminals(ReconnectionType) || []) {
+			const reconnectionData = terminal.shellLaunchConfig.attachPersistentProcess?.reconnectionProperties?.data as IReconnectionTaskData;
+			if (reconnectionData) {
+				this._reconnectedTerminals.push({ ...terminal, reconnectionData });
+			} else {
+				this._logService.trace(`Terminal was owned by tasks but has no reconnection data ${terminal.instanceId}, ${JSON.stringify(terminal.shellLaunchConfig)}`);
```
Comment (meganrogge) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1335]: I'm not sure how this was possible before 🤔

Comment (meganrogge) [reply src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1335]: which makes me think this shouldn't have ever been undefined

https://github.com/microsoft/vscode/pull/182896/files#diff-5d6d50fef79440567acc5f9a977c908f45f59342bb4d5e19ef5b9ee25e02920cL1280

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 2]
```diff
@@ -1277,8 +1277,8 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 		}
 		for (let i = 0; i < this._reconnectedTerminals.length; i++) {
 			const terminal = this._reconnectedTerminals[i];
-			const taskForTerminal = terminal.shellLaunchConfig.attachPersistentProcess?.reconnectionProperties?.data as IReconnectionTaskData;
-			if (taskForTerminal.lastTask === task.getCommonTaskId()) {
+			const data = terminal.shellLaunchConfig?.attachPersistentProcess?.reconnectionProperties?.data as IReconnectionTaskData | undefined;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1280]: `shellLaunchConfig` should always exist

```suggestion
			const data = terminal.shellLaunchConfig.attachPersistentProcess?.reconnectionProperties?.data as IReconnectionTaskData | undefined;
```

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 3]
```diff
@@ -1323,19 +1323,17 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
...
 		this._logService.trace(`Attempting reconnection of ${this._reconnectedTerminals?.length} terminals`);
 		if (!this._reconnectedTerminals?.length) {
 			this._logService.trace(`No terminals to reconnect to so returning`);
 		} else {
 			for (const terminal of this._reconnectedTerminals) {
-				const task = terminal.shellLaunchConfig.attachPersistentProcess?.reconnectionProperties?.data as IReconnectionTaskData;
-				this._logService.trace(`Reconnecting to task: ${JSON.stringify(task)}`);
-				if (!task) {
-					continue;
+				const data = terminal.shellLaunchConfig?.attachPersistentProcess?.reconnectionProperties?.data as IReconnectionTaskData | undefined;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1332]:
```suggestion
				const data = terminal.shellLaunchConfig.attachPersistentProcess?.reconnectionProperties?.data as IReconnectionTaskData | undefined;
```

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts (hunk 2 above)
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1280]: You access .data a couple of times, should it move to a helper method?

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 4]
```diff
@@ -1323,19 +1323,17 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
...
+				if (data) {
+					const terminalData = { lastTask: data.lastTask, group: data.group, terminal };
+					this._terminals[terminal.instanceId] = terminalData;
 				}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1336]: Do we need a log here?

**Review by Tyriar — CHANGES_REQUESTED** [review]

**Review by Tyriar — APPROVED** [review]

**Pull request merged**